### PR TITLE
Add l1 and min split loss regularization parameters

### PR DESCRIPTION
**Pull request by RAMitchell**
Implements #203

The parameter names might still be changed.

### CODE REVIEW

**Comment by RAMitchell** [comment]
I've moved away from greek letters in naming regularization parameters. It causes confusion when legate-boost/sklearn/xgboost are all using the same letter to mean different things (e.g. xgb alpha is l1, sklearn alpha is l2). There is no consistent convention.

File: legateboost/models/tree.py [hunk 1]
```diff
@@ -32,8 +33,15 @@ class Tree(BaseModel):
     split_samples : int
         The number of data points to sample for each split decision.
         Max value is 2048 due to constraints on shared memory in GPU kernels.
-    alpha : float
-        The L2 regularization parameter.
+    l1_regularization : float
+        The L1 regularization parameter applied to leaf weights.
+    l2 : float
```
Comment (seberg) [inline legateboost/models/tree.py:38]:
```suggestion
    l2_regularization : float
```

File: legateboost/models/tree.py [hunk 2]
```diff
@@ -46,13 +54,27 @@ def __init__(
         self,
         max_depth: int = 8,
         split_samples: int = 256,
-        alpha: float = 1.0,
```
Comment (seberg) [inline legateboost/models/tree.py:49]: FWIW, I think it would make sense to use a `, *,` here (strictly speaking necessary to since passing by position now makes it the `l1_regularization` parameter). 

(Didn't revisit previous signatures)

And/or maybe just put `l2_reg` first always. since it is also the pattern in the C-code.

File: legateboost/models/tree.py [hunk 3]
```diff
@@ -46,13 +54,27 @@ def __init__(
...
+        l1_regularization: float = 0.0,
+        l2_regularization: float = 1.0,
```
Comment (seberg) [inline legateboost/models/tree.py:58]: It is nice to have the actual constant here.  But is it normal that if you use `l1` regularization you *also* use `l2` regularization?

Because if not, it seems better to me to default to `None` on both set `l2_regularization = 1.0` only if both l1 and l2 are `None`.

Otherwise it seems an easy mistake to pass `l1_regularization=1.0` and not realize that l2 is also set.

File: legateboost/legateboost.py [hunk 4]
```diff
@@ -241,6 +241,8 @@ def _partial_fit(
         validate_data(self, X, reset=False, skip_check_array=True)
         _eval_set = self._process_eval_set(eval_set)
         sample_weight = check_sample_weight(sample_weight, y.shape[0])
+        if not isinstance(self.base_models, tuple) or len(self.base_models) == 0:
+            raise ValueError("base_models should be a tuple")
```
Comment (seberg) [inline legateboost/legateboost.py:245]: Would make sense to add the `== 0` check also in the class instantiation?  Unless it is expected to be mutated at that point.

File: src/models/tree/build_tree.cc [hunk 5]
```diff
@@ -298,7 +298,12 @@ struct TreeBuilder {
       }
     }
   }
-  void PerformBestSplit(Tree& tree, Histogram<GPair> histogram, double alpha, NodeBatch batch)
+  void PerformBestSplit(Tree& tree,
+                        Histogram<GPair> histogram,
+                        double l2_regularization,
+                        NodeBatch batch,
+                        double l1_regularization,
```
Comment (seberg) [inline src/models/tree/build_tree.cc:305]:
```suggestion
                        NodeBatch batch,
                        double l2_regularization,
                        double l1_regularization,
```
maybe always l2 + l1 together (in that order)

File: legateboost/models/tree.py [hunk 6]
```diff
@@ -74,10 +96,12 @@ def fit(
         task.add_scalar_arg(self.max_depth, types.int32)
         max_nodes = 2 ** (self.max_depth + 1)
         task.add_scalar_arg(max_nodes, types.int32)
-        task.add_scalar_arg(self.alpha, types.float64)
+        task.add_scalar_arg(self.l2_regularization, types.float64)
```
Comment (seberg) [inline legateboost/models/tree.py:99]: Can you move this down to be before the l1 regularization?  (or move the l1 up here)

File: src/models/tree/build_tree.h [hunk 7]
```diff
@@ -102,9 +103,50 @@ inline __host__ __device__ auto ComputeHistogramBin(int node_id,
...
+}  // namespace detail
+
+__host__ __device__ inline auto CalculateGain(const GPair& left_sum,
+                                              const GPair& right_sum,
+                                              const GPair& parent_sum,
+                                              double alpha,
+                                              double l1_regularization,
+                                              double gamma) -> double
+{
+  double const reg = std::max(1e-5, alpha);
```
Comment (seberg) [inline src/models/tree/build_tree.h:130]: Why not keep using `eps`?

File: src/models/tree/build_tree.h [hunk 8]
```diff
@@ -102,9 +103,50 @@ inline __host__ __device__ auto ComputeHistogramBin(int node_id,
...
+  }
+  // Note: we could use the cuda __ddiv_rn intrinsic here if this becomes a bottleneck
+  return (z * z) / (sum.hess + alpha_reg);
+}
+}  // namespace detail
+
+__host__ __device__ inline auto CalculateGain(const GPair& left_sum,
+                                              const GPair& right_sum,
+                                              const GPair& parent_sum,
+                                              double alpha,
```
Comment (seberg) [inline src/models/tree/build_tree.h:126]: Maybe after the `std::max` passing it as `alpha` makes sense maybe, but I think this should be `l2_regularization` still.

File: src/models/tree/build_tree.h [hunk 9]
```diff
@@ -102,9 +103,50 @@ inline __host__ __device__ auto ComputeHistogramBin(int node_id,
...
-  return -G / (H + alpha);
+  // l1 threshold
+  double z = 0.0;
+  if (sum.grad > l1_regularization) {
+    z = sum.grad - l1_regularization;
+  } else if (sum.grad < -l1_regularization) {
+    z = sum.grad + l1_regularization;
+  }
+  // Note: we could use the cuda __ddiv_rn intrinsic here if this becomes a bottleneck
+  return (z * z) / (sum.hess + alpha_reg);
```
Comment (seberg) [inline src/models/tree/build_tree.h:119]: We could be tricky and use that `z * z` is always positive, but likely the math optimizer is able to even see that.

File: src/models/tree/build_tree.h [hunk 10]
```diff
@@ -102,9 +103,50 @@ inline __host__ __device__ auto ComputeHistogramBin(int node_id,
...
+}
+}  // namespace detail
+
+__host__ __device__ inline auto CalculateGain(const GPair& left_sum,
+                                              const GPair& right_sum,
+                                              const GPair& parent_sum,
+                                              double alpha,
+                                              double l1_regularization,
+                                              double gamma) -> double
```
Comment (seberg) [inline src/models/tree/build_tree.h:128]: Should we keep with `min_split_gain` here?

File: legateboost/models/tree.py [hunk 11]
```diff
@@ -32,8 +33,15 @@ class Tree(BaseModel):
...
-        The L2 regularization parameter.
+    l1_regularization : float
+        The L1 regularization parameter applied to leaf weights.
+    l2 : float
+        The L2 regularization parameter applied to leaf weights.
+    alpha : deprecated
+        Deprecated, use `l2_regularization` instead.
+    min_split_gain : float
+        The minimum improvement in the loss function required to make a split.
+        Increasing this value generates smaller trees.
```
Comment (seberg) [inline legateboost/models/tree.py:44]: As discussed, may make sense to note that this is an improvement per number of classes.

**Review by seberg — COMMENTED** [review]
Thanks, will admit to reviewing this from a "changes" perspective, but I really like the refactor into helper functions.

Looks great, and only 1-2 small nits to clarify/change!

Most comments are take-or-leave as always :).  I think it would make sense to just start using `*,` for args.  I would use `DeprecationWarning`, but I think this is all good and pragmatic.

I think besides small style cleanups, the only real thing is whether we shouldn't default to `l2_reg=0.` if `l1_reg=1.` is passed.

File: src/models/tree/build_tree.cc (hunk 5 above)
Comment (RAMitchell) [reply src/models/tree/build_tree.cc:305]: I'll switch it to l1, l2.

File: src/models/tree/build_tree.h (hunk 7 above)
Comment (RAMitchell) [reply src/models/tree/build_tree.h:130]: I had a compiler error where you cannot take a reference to static const variable in device code - std::max takes a reference. Its fixed by doing `double const reg = std::max(double(eps), alpha);` to force a copy.

File: src/models/tree/build_tree.h (hunk 9 above)
Comment (RAMitchell) [reply src/models/tree/build_tree.h:119]: I don't follow. Is there a way to do faster division if you know the numerator is positive?

Comment (seberg) [reply src/models/tree/build_tree.h]: sorry, I was just thinking wrong, the comment was nonsense.

File: legateboost/models/tree.py (hunk 3 above)
Comment (seberg) [reply legateboost/models/tree.py:61]: OK, thanks.  Still find this a bit surprising, but I now agree this seems to be what XGBoost does, so it's fine.

**Review by seberg — APPROVED** [review]
Thanks for the polishing and answering my questions :).
